Allow user to rebuild incompatible packages against the current version of Atom

nathansobo opened the pull request:
This PR depends on core APIs introduced in https://github.com/atom/atom/pull/8475.

![rebuild-packages](https://cloud.githubusercontent.com/assets/1789/9421740/47d8e454-4836-11e5-8445-9536ec2fa846.gif)

### Tasks

* [x] Remove extra text
* [x] Add rebuild all button
* [x] Disable rebuild button when package rebuild starts
* [x] Indicate rebuild in progress for each package
* [x] When build succeeds, indicate it
* [x] When build fails, indicate it with error messages
* [x] Persist build failure and don't load packages that failed to build (atom/atom PR pending)
* [x] Add button linking to package settings
* [x] When all rebuilds complete, show notification with a reload button

### Testing Tips

The easiest way to make an incompatible package is to take a package that has native dependencies and run `npm rebuild` to compile it against the Node headers. I use the following:

* `spell-check`
* `dev-live-reload` Make sure you're checked out to `0.46.0`.

For both these packages do the following:

* `rm -rf node_modules`
* `apm install`
* `npm rebuild`
* `apm link -d`

Now both should show up in your incompatible packages listing in dev mode.

To create a build failure, add some random syntax error in `spell-check/node_modules/spellchecker/src/main.cc` before rebuilding. Be aware: The rebuild failure will be persisted until the next time you click rebuild.

kevinsawicki commented:
Awesome, thanks so much for getting this going :+1:

nathansobo commented:
@simurai Heads up... I'm going to do this kinda ugly and ask for you to swoop in with some design sauce. I have a decent amount of the logic in place. The basic flow is as follows:

* If there are no incompatible packages, user is presented with a message saying this.
* If there are incompatible packages, user is told this and there's a rebuild button.
* The user is presented with all packages containing incompatible node modules, with the individual modules listed along with the reason they are incompatible.
* When the user clicks rebuild all, each package rebuilds in sequence... we need to indicate this is in progress.
* When a package rebuilds cleanly, we need to indicate this somehow.
* When a package fails to rebuild, we need to show an error message.

Also, if you whip up a mockup of some sort I could also just implement it.

simurai commented:
@nathansobo I think you already started. I can brush over it on Mon/Tue. You don't have to worry about styling as long as it works.

nathansobo commented:
I added an animated GIF to first comment to illustrate today's progress. Almost there except for the interaction following build failure.

@thomasjo If you want to work on adding back the buttons to uninstall upgrade and create issues, that would be :basketball:er. I'm thinking we shouldn't display the create issue button unless we know the user is up to date... I updated the checklist accordingly and added some hints about testing this in dev mode to the body.

nathansobo commented:
Oh, I almost forgot, this depends on APIs introduced in https://github.com/atom/atom/pull/8475.

thomasjo commented:
Kinda hit a roadblock since there's no API for either uninstalling packages or checking for updates. All of that stuff is located in the settings-view (and apm of course). Do we simply want to copy and paste the relevant code from settings-view, or should we expose it somehow?

Current behavior is simply to offer ability to check for updates, report issue, and disable the package.

thomasjo commented:
How do we want to deal with linked packages? I'm assuming we won't check for updates for those packages, so should we always display a *Report Issue* button in that case?

thomasjo commented on lib/incompatible-packages-component.js:
```diff
@@ -0,0 +1,234 @@
+      )
+    }
+  }
+
```
@nathansobo :fire: This method? It's no longer being used.

nathansobo commented:
@thomasjo Okay, I just added a link to the package settings.

@simurai This is ready for some :art: from you. You can follow the steps listed in the body to get it into different states. My main concern is the relationship between the settings button and the badge.

simurai commented:
@nathansobo Got `spell-check` into an incompatible state. But having problems with this branch. I did:

1. Checkout `incompatible-packages/tree/ns-mb-rebuild`
2. Checkout `atom/tree/ns-rebuild-packages`
3. Run `script/build`. First time got an error. Second time it worked.
4. Open Atom. Now I get this error:

![screen shot 2015-08-25 at 8 56 50 pm](https://cloud.githubusercontent.com/assets/378023/9466055/b1a3778a-4b6c-11e5-9fe6-fbb042dc0f94.png)

```
A JavaScript error occurred in the main process

Uncaught Exception:
Error: Module version mismatch. Expected 44, got 43.
    at Error (native)
    at Object.module.(anonymous function) (ATOM_SHELL_ASAR.js:137:20)
    at Object.module.(anonymous function) [as .node] (ATOM_SHELL_ASAR.js:148:18)
    at Module.load (module.js:353:32)
    at Function.Module._load (module.js:308:12)
    at Module.require (module.js:363:17)
    at require (module.js:382:17)
    at Object.<anonymous> (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/nslog/lib/nslog.js:5:11)
    at Object.<anonymous> (/Applications/Atom.app/Contents/Resources/app.asar/node_modules/nslog/lib/nslog.js:15:4)
    at Module._compile (module.js:428:26)
```

Did I miss something else? It looks the same as @lee-dohm's error here: https://github.com/atom/atom/pull/7877#issuecomment-127274105

nathansobo commented:
@simurai Ah shit I rebased on top of the Electron upgrade for testing everything combined.

simurai commented:
@nathansobo Will the Electron upgrade land in master soon? Or will this PR go first? I can try to rebase too, although I somehow already messed up Atom with switching branches.

> My main concern is the relationship between the settings button and the badge.

Based on the gif.. ya, the badges might look a bit like they could be buttons and not just some status indicator. Maybe the badges could only be used for the icons and not include the text. But that can also be changed later, after this shipped.

nathansobo commented:
Electron upgrade will be landing very soon. Let's just wait until then to proceed with styling changes.

nathansobo commented:
The build fails against the current release, but we need to bundle this to cut a new release, so I'm just going to merge due to this passing locally.
